pkg/_internal/gracePeriod/gracePeriod.go

```diff
@@ -0,0 +1,57 @@
+package gracePeriod
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/_internal/metadata"
+)
+
+const (
+	graceTimestampAnnotation = "kuadrant.io/grace-timeout"
+)
+
+var ErrGracePeriodNotExpired = fmt.Errorf("grace period has not yet expired")
+
+func WasGracePeriodNotExpiredErr(e error) bool {
+	return strings.Contains(e.Error(), "grace period has not yet expired")
+}
+
+func GracefulDelete(ctx context.Context, c client.Client, obj client.Object, at time.Time) error {
+	if err := c.Get(ctx, client.ObjectKeyFromObject(obj), obj); err != nil {
+		return err
+	}
+
+	// if At is before the current time, just delete it
+	if at.Before(time.Now()) || at.Equal(time.Now()) {
+		return c.Delete(ctx, obj)
+	}
+
+	//ensure finalizer and annotation are present
+	if !metadata.HasAnnotation(obj, graceTimestampAnnotation) {
```

Comment:
is this always a "downstream" object? I find it weird that we are setting the annotation